services/src/test/scala/cromwell/pipeline/service/ProjectServiceTestImpl.scala

```diff
@@ -0,0 +1,75 @@
+import cromwell.pipeline.model.wrapper.UserId
+import cromwell.pipeline.service.ProjectService.Exceptions.{ ProjectAccessDeniedException, ProjectNotFoundException }
+import cromwell.pipeline.service.VersioningException.RepositoryException
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+class ProjectServiceTestImpl(projectRepository: ProjectRepository, testMode: ServiceTestMode)(
+  implicit executionContext: ExecutionContext
+) extends ProjectService {
+
+  private[service] def getUserProjectById(projectId: ProjectId, userId: UserId): Future[Project] =
+    projectRepository.getProjectById(projectId).flatMap(project => getForUserOrFail(project.toSeq, userId))
+
+  def getUserProjectByName(namePattern: String, userId: UserId): Future[Project] =
+    testMode match {
+      case WithException => Future.failed(new RuntimeException("Something went wrong"))
+      case _             => projectRepository.getProjectsByName(namePattern).flatMap(getForUserOrFail(_, userId))
+    }
+
+  private def getForUserOrFail(projects: Seq[Project], userId: UserId): Future[Project] =
+    projects.find(_.ownerId == userId) match {
+      case Some(project)             => Future.successful(project)
+      case None if projects.nonEmpty => Future.failed(new ProjectAccessDeniedException)
+      case _                         => Future.failed(new ProjectNotFoundException)
+    }
+
+  def addProject(request: ProjectAdditionRequest, userId: UserId): Future[Either[VersioningException, Project]] =
+    testMode match {
+      case WithException => Future.successful(Left(RepositoryException("VersioningException.RepositoryException")))
+      case _             => Future.successful(Right(TestProjectUtils.getDummyProject(name = request.name, ownerId = userId)))
+    }
+
+  def deactivateProjectById(projectId: ProjectId, userId: UserId): Future[Project] =
+    testMode match {
+      case WithException => Future.failed(new RuntimeException("Something wrong"))
+      case _             => projectRepository.deactivateProjectById(projectId).flatMap(_ => getUserProjectById(projectId, userId))
+    }
+
+  def updateProjectName(projectId: ProjectId, request: ProjectUpdateNameRequest, userId: UserId): Future[ProjectId] =
+    testMode match {
+      case WithException => Future.failed(new RuntimeException("Something wrong"))
+      case _ =>
+        getUserProjectById(projectId, userId).flatMap { project =>
+          val updatedProject = project.copy(name = request.name)
+          projectRepository.updateProjectName(updatedProject).map(_ => updatedProject.projectId)
+        }
+    }
+
+  def updateProjectVersion(projectId: ProjectId, version: PipelineVersion, userId: UserId): Future[Int] =
+    testMode match {
+      case WithException => Future.failed(new RuntimeException("Something wrong"))
+      case _ =>
+        getUserProjectById(projectId, userId).flatMap { project =>
+          projectRepository.updateProjectVersion(project.copy(version = version))
+        }
```

Comment:
Do we need such checks in test implementation? Can't we just use `projectRepository.updateProjectVersion(project.copy(version = version))` ? Same for `getUserProjectByName`, `updateProjectName`. Just update / return it or fail with error from `WithException`

Reply:
Refactored that class, `ProjectRepository` dependency was removed. Can that approach be better here? Or should I rollback to using instance of `ProjectRepositoryTestImpl`